File: cbta97net-fix-editar-alumno/cbta97net-front/src/app/nucleo/servicios/estructura/materia.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from 'environments/environment';
import { Materia } from '@modelos/comunes/materia.model';

@Injectable({
    providedIn: 'root'
})
export class MateriaServicio {

    private readonly URL_MATERIAS = `${environment.apiUrl}/materias`;

    /**
     * Constructor por defecto. Permite asignar clientes HTTP mockeables para
     * realizacion de pruebas.
     * @param httpClient cliente para realizar peticiones HTTP.
     */
    constructor(private readonly http: HttpClient) {}

    /**
     * Obtiene una materia específica mediante su identificador único.
     * @param id El ID de la materia a buscar.
     * @returns Un Observable con el objeto Materia encontrado.
     */
    obtenerMateriaPorId(id: number): Observable<Materia> {
        return this.http.get<Materia>(`${this.URL_MATERIAS}/${id}`);
    }

    /**
     * Registra una nueva materia en el sistema.
     * @param materia Objeto con los datos de la materia a guardar.
     * @returns Un Observable con la materia creada y su ID asignado.
     */
    registrarMateria(materia: Materia): Observable<Materia> {
        return this.http.post<Materia>(this.URL_MATERIAS, materia);
    }

    /**
     * Recupera el listado completo de todas las materias registradas.
     * @returns Un Observable con un arreglo de objetos Materia.
     */
    obtenerTodasLasMaterias(): Observable<Materia[]> {
        return this.http.get<Materia[]>(this.URL_MATERIAS);
    }

    /**
     * Filtra y obtiene las materias pertenecientes a una carrera técnica específica.
     * @param carreraTecnicaId ID de la carrera técnica.
     * @returns Un Observable con el listado de materias de dicha carrera.
     */
    obtenerMateriasPorCarrera(carreraTecnicaId: number): Observable<Materia[]> {
        return this.http.get<Materia[]>(`${this.URL_MATERIAS}/carrera/${carreraTecnicaId}`);
    }

    /**
     * Recupera las materias asociadas a un área propedéutica específica.
     * @param id ID del área propedéutica.
     * @returns Un Observable con el listado de materias filtradas por área.
     */
    obtenerMateriasPorAreaPropedeutica(id: number): Observable<Materia[]> {
        return this.http.get<Materia[]>(`${this.URL_MATERIAS}/area/${id}`);
    }

    /**
     * Obtiene las materias correspondientes a un grado o semestre específico.
     * @param grado Número del semestre (Ej. 1 al 6).
     * @returns Un Observable con las materias impartidas en ese semestre.
     */
    obtenerMateriasPorGrado(grado: number): Observable<Materia[]> {
        return this.http.get<Materia[]>(`${this.URL_MATERIAS}/grado/${grado}`);
    }

    /**
     * Realiza una búsqueda filtrando simultáneamente por grado y carrera técnica.
     * @param grado Número del semestre.
     * @param carreraTecnicaId ID de la carrera técnica.
     * @returns Un Observable con las materias que coinciden con ambos criterios.
     */
    obtenerMateriasPorGradoYCarrera(grado: number, carreraTecnicaId: number): Observable<Materia[]> {
        return this.http.get<Materia[]>(`${this.URL_MATERIAS}/grado/${grado}/carrera/${carreraTecnicaId}`);
    }

    /**
     * Búsqueda avanzada de materias aplicando tres filtros: grado, carrera y área propedéutica.
     * @param grado Número del semestre.
     * @param carreraId ID de la carrera técnica.
     * @param areaId ID del área propedéutica.
     * @returns Un Observable con las materias que cumplen con los tres parámetros de búsqueda.
     */
    obtenerMateriasCompleto(grado: number, carreraId: number, areaId: number): Observable<Materia[]> {
        return this.http.get<Materia[]>(`${this.URL_MATERIAS}/grado/${grado}/carrera/${carreraId}/area/${areaId}`);
    }
}

File: cbta97net-fix-editar-alumno/cbta97net-backend/src/main/java/mx/edu/cbta/sistemaescolar/estructura/service/MateriaService.java
package mx.edu.cbta.sistemaescolar.estructura.service;

import mx.edu.cbta.sistemaescolar.estructura.domain.exception.ImportarMateriasSISEEMSException;
import mx.edu.cbta.sistemaescolar.estructura.domain.exception.SemestreMateriaNoValidoException;
import mx.edu.cbta.sistemaescolar.estructura.domain.exception.MateriaNoEncontradaException;
import mx.edu.cbta.sistemaescolar.estructura.domain.exception.MateriaDuplicadaException;
import mx.edu.cbta.sistemaescolar.estructura.dto.MateriaDTO;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Interfaz de servicio para la gestión de Materias/Asignaturas.
 * Administra el currículo académico y las relaciones jerárquicas entre materias.
 */
public interface MateriaService {

    /**
     * Consulta una materia específica por su identificador.
     * @param id ID de la materia.
     * @return {@link MateriaDTO} con los datos de la asignatura.
     * @throws MateriaNoEncontradaException Si el registro no existe.
     */
    MateriaDTO obtenerMateriaPorId(Long id) throws MateriaNoEncontradaException;

    /**
     * Registra una nueva asignatura en el catálogo curricular.
     * @param materiaDto Datos de la asignatura a registrar.
     * @return {@link MateriaDTO} persistido.
     * @throws SemestreMateriaNoValidoException Si el semestre asignado es fuera de rango.
     * @throws MateriaDuplicadaException Si la clave o nombre ya existe.
     */
    MateriaDTO registrarMateria(MateriaDTO materiaDto) throws SemestreMateriaNoValidoException, MateriaDuplicadaException;

    /**
     * Obtiene todas las materias sin filtros aplicados.
     * @return Lista total de {@link MateriaDTO}.
     */
    List<MateriaDTO> obtenerTodasLasMaterias();

    /**
     * Filtra las materias pertenecientes a un Área Propedéutica específica.
     * @param areaPropedeuticaId ID del área de especialización.
     * @return Lista de materias del área indicada.
     */
    List<MateriaDTO> obtenerTodasPorAreaPropedeutica(Long areaPropedeuticaId);

    /**
     * Obtiene las materias asociadas a una Carrera Técnica.
     * @param carreraTecnicaId ID de la carrera técnica.
     * @return Lista de materias curriculares de la carrera.
     */
    List<MateriaDTO> obtenerMateriasPorCarrera(Long carreraTecnicaId);

    /**
     * Consulta las materias que se imparten en un semestre específico.
     * @param semestre Número del semestre (1-6).
     * @return Lista de materias del nivel solicitado.
     */
    List<MateriaDTO> obtenerMateriasPorSemestre(int semestre);

    /**
     * Filtra materias por semestre y por carrera técnica asociada.
     * @param semestre Nivel académico solicitado.
     * @param carreraTecnicaId Carrera técnica de interés.
     * @return Lista de materias que cumplen ambos criterios.
     */
    List<MateriaDTO> obtenerMateriasPorSemestreYCarrera(int semestre, Long carreraTecnicaId);

    /**
     * Realiza una búsqueda exhaustiva por semestre, carrera y área propedéutica.
     * @param semestre Semestre a consultar.
     * @param carreraTecnicaId Identificador de la carrera.
     * @param areaPropedeuticaId Identificador del área.
     * @return Lista de materias filtrada por los tres criterios.
     */
    List<MateriaDTO> obtenerMateriasPorSemestreYCarreraYArea(int semestre, Long carreraTecnicaId, Long areaPropedeuticaId);

    /**
     * Importa las materias presentes en el archivo provisto por SISEEMS y las almacena
     * en el sistema.
     * Las materias ya existentes en el sistema se omiten al realizar el registro
     * masivo.
     * @param archivoSISEEMS Archivo de tipo Hoja de Calculo (Excel) de SISEEMS.
     * @throws ImportarMateriasSISEEMSException Si ocurre un error al intentar obtener los archivos del archivo.
     */
    void importarMateriasSISEEMS(MultipartFile archivoSISEEMS) throws ImportarMateriasSISEEMSException;
}